diff --git a/include/linux/sched.h b/include/linux/sched.h
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -553,6 +553,23 @@ struct sched_entity {
 	u64				nr_migrations;
 
 	struct sched_statistics		stats;
+#ifdef CONFIG_SCHED_BORE
+	struct sched_burst_cache {
+		u8		score;
+		u32		count;
+		u64		timestamp;
+		spinlock_t	lock;
+	} child_burst;
+	struct sched_burst_cache group_burst;
+	u64				burst_time;
+	u8				prev_burst_penalty;
+	u8				curr_burst_penalty;
+	u8				burst_penalty;
+	u8				burst_score;
+#endif
+
 #ifdef CONFIG_SMP
 	u32				on_rq;
 #endif