--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -125,6 +125,12 @@
 unsigned int sysctl_sched_min_granularity			= 750000ULL;
 unsigned int sysctl_sched_idle_min_granularity			= 750000ULL;
 
+/* BORE: Priority for Interactive Tasks */
+unsigned int sysctl_sched_bore = 1;
+unsigned int sysctl_sched_burst_fork_at_max = 1;
+unsigned int sysctl_sched_burst_penalty_scale = 1280;
+unsigned int sysctl_sched_burst_penalty_offset = 0;
+
 /*
  * after partition, the default both 0.5ms
  */